File: java/src/leetcode/editor/cn/98-ValidateBinarySearchTree.java
package leetcode.editor.cn;

import ivonhoe.java.leetcode.bean.TreeNode;

/*
 * @lc app=leetcode.cn id=98 lang=java
 *
 * [98] Validate Binary Search Tree
 *
 * https://leetcode-cn.com/problems/validate-binary-search-tree/description/
 *
 * algorithms
 * Medium (35.57%)
 * Likes:    1835
 * Dislikes: 0
 * Total Accepted:    643K
 * Total Submissions: 1.8M
 * Testcase Example:  '[2,1,3]'
 *
 * Given the root of a binary tree, determine if it is a valid binary search
 * tree (BST).
 * 
 * A valid BST is defined as follows:
 * 
 * 
 * The left subtree of a node contains only nodes with keys less than the
 * node's key.
 * The right subtree of a node contains only nodes with keys greater than the
 * node's key.
 * Both the left and right subtrees must also be binary search trees.
 * 
 * 
 * 
 * Example 1:
 * 
 * 
 * Input: root = [2,1,3]
 * Output: true
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: root = [5,1,4,null,null,3,6]
 * Output: false
 * Explanation: The root node's value is 5 but its right child's value is
 * 4.
 * 
 * 
 * 
 * Constraints:
 * 
 * 
 * The number of nodes in the tree is in the range [1, 10^4].
 * -2^31 <= Node.val <= 2^31 - 1
 * 
 * 
 */

// @lc code=start
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode() {} TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) { this.val = val; this.left
 * = left; this.right = right; } }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        // isValidBST(root, null, null);
    }

    // 中序遍历
    long pre = Long.MIN_VALUE;

    private boolean inorderValidBST(TreeNode root) {
        if (root == null) {
            return true;
        }

        if (!inorderValidBST(root.left)) {
            return false;
        }

        if (root.val <= pre) {
            return false;
        }

        pre = root.val;
        return inorderValidBST(root.right);
    }

    // 先序遍历
    private boolean isValidBST(TreeNode root, TreeNode min, TreeNode max) {
        if (root == null) {
            return true;
        }
        if (max != null && root.val >= max.val) {
            return false;
        }
        if (min != null && root.val <= min.val) {
            return false;
        }

        return isValidBST(root.left, min, root) && isValidBST(root.right, root, max);
    }
}
// @lc code=end

File: kotlin/src/leetcode/editor/cn/98-ValidateBinarySearchTree.kt
import leetcode.editor.bean.TreeNode

//给定一个二叉树，判断其是否是一个有效的二叉搜索树。
//
// 假设一个二叉搜索树具有如下特征： 
//
// 
// 节点的左子树只包含小于当前节点的数。 
// 节点的右子树只包含大于当前节点的数。 
// 所有左子树和右子树自身必须也是二叉搜索树。 
// 
//
// 示例 1: 
//
// 输入:
//    2
//   / \
//  1   3
//输出: true
// 
//
// 示例 2: 
//
// 输入:
//    5
//   / \
//  1   4
//     / \
//    3   6
//输出: false
//解释: 输入为: [5,1,4,null,null,3,6]。
//     根节点的值为 5 ，但是其右子节点值为 4 。
// 
// Related Topics 树 深度优先搜索 递归 
// 👍 920 👎 0


//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {

    fun isValidBST(root: TreeNode?): Boolean {
        return solution2(root, null, null)
    }

    // 非递归实现
//    fun solution2(root: TreeNode?): Boolean {
//        var stack = Stack()
//
//        return false
//    }

    // 递归实现
    // 二叉搜索树，中序遍历是升序的，所以检查前一个节点和当前节点的大小
    var pre: Long = Long.MIN_VALUE
    fun solution1(root: TreeNode?): Boolean {
        if (root?.`val` == null) {
            return true
        }
        if (!solution1(root.left)) {
            return false
        }
        if (root.`val` <= pre) {
            return false
        }
        pre = root.`val`.toLong()

        return solution1(root.right)
    }

    // 前序遍历
    fun solution2(root: TreeNode?, min: TreeNode?, max: TreeNode?): Boolean {
        if (root == null) {
            return true
        }
        if (min != null && root.`val` <= min.`val`) {
            return false
        }
        if (max != null && root.`val` >= max.`val`) {
            return false
        }
        return solution2(root.left, min, root) && solution2(root.right, root, max)
    }
}
//leetcode submit region end(Prohibit modification and deletion)
